### ISSUE TITLE
[reflection] Get list of all available definitions?

### ISSUE THREAD

**Issue by t-more** [agda/agda#4581]
I've been working a bit on some rewriting tactics for Agda. I would like to create a tactic that pulls known equality proofs from the local file etc, but I seem to be unable to find any functions that returns such. Does anyone know how to do it? Alternatively, can I make my own functions on the TC monad using the Haskell FFI?

**Comment by jespercockx** [comment]
It is not possible with the current reflection infrastructure to get a list of all definitions in the current file/module. I don't think it would be too hard to add such a thing, but I guess the performance could potentially be quite bad when you have a lot of imported modules. So perhaps the better question is if we can implement some kind of "hint databases" that can be used by tactics to store and retrieve certain lemmas based on their name, type, etc.

**Comment by uhbif19** [comment]
@jespercockx Should not such "hint database" usecase be covered by some "Hint" typeclass instances?

**Comment by jespercockx** [comment]
Yes, you can now use the `getInstances` TC primitive to get all the instances that match a particular type. This does not match the use case of the OP precisely because (1) it requires definitions to be marked as instances and (2) you cannot choose to only get instances from a particular module. Still, it might be sufficient for much of the same use cases.

**Comment by uhbif19** [comment]
Yes, but both of this points does not seem to interfere with "hint database" idea. Hint typeclass might have additional field(s) for selecting applicable hints, I think this is similar to how hints work in Coq.

This probably does not cover case of writing auto-like tactic, though.
